File: 00200-valid-parentheses/main.go
package main

import (
	"fmt"

	"github.com/emirpasic/gods/stacks/linkedliststack"
)

func isValid(s string) bool {
	stack := linkedliststack.New()
	for _, c := range s {
		if c == '(' || c == '[' || c == '{' {
			stack.Push(c)
		} else {
			top, ok := stack.Pop()
			if !ok {
				return false
			}

			if c == ')' && top != '(' {
				return false
			}
			if c == ']' && top != '[' {
				return false
			}
			if c == '}' && top != '{' {
				return false
			}
		}
	}
	return stack.Empty()
}

func main() {
	fmt.Println(isValid("()"))
	fmt.Println(isValid("()[]{}"))
	fmt.Println(isValid("(]"))
}

File: 00200-valid-parentheses/main.js
function isValid(s) {
  const stack = [];
  // console.log("isValid", s);
  for (let i = 0; i < s.length; i++) {
    // console.log("Char at", i, s[i]);
    if (s[i] == "(" || s[i] == "[" || s[i] == "{") {
      stack.push(s[i]);
    } else {
      if (stack.length == 0) {
        return false;
      }
      const top = stack.pop();
      if (top == "[" && s[i] != "]") {
        return false;
      }
      if (top == "(" && s[i] != ")") {
        return false;
      }
      if (top == "{" && s[i] != "}") {
        return false;
      }
    }
  }

  return stack.length == 0;
}

s = "()";
console.log("Example 1", isValid(s));

s = "()[]{}";
console.log("Example 2", isValid(s));

s = "(]";
console.log("Exampel 3", isValid(s));

s = "[";
console.log("Exampel 4", isValid(s));
